2310 Add load_csv_datalist utility API

Nic-Ma opened the pull request:
Fixes #2310 .

### Description
This PR added the `load_csv_datalist` API to load extra information from CSV files.
Users can easily combine this datalist with the `image` and `label`, etc. and put in `Dataset`.

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

Nic-Ma commented:
Thanks for Seyed's example code and use cases, I totally changed my previous local code and switched to `pandas`.
Will complete this PR as soon as possible.

Thanks.

Nic-Ma commented:
/black

Nic-Ma commented:
/black

Nic-Ma commented:
Hi @wyli @ericspod ,

This PR is ready for review, I will also develop a tutorial to show how to use it with `load_decathlon_datalist()` when this PR merged.

Thanks.

wyli commented:
I think the CSV reading should be implemented with the MONAI dataset API, with an option of partially loading large csv files e.g. https://discuss.pytorch.org/t/how-to-use-dataset-larger-than-memory/37785

Nic-Ma commented:
Hi @wyli ,

Thanks for your suggestion.
I will investigate how to build a Dataset above this CSV utility and support partially loading.

Thanks.

Nic-Ma commented:
Hi @wyli ,

I want to double confirm your suggestion: if we partially load a large CSV in dataset, do you mean to only load chunks of the CSV for a training, or still use the whole dataset for training but don't load data before training, everytime we only open the CSV file to **read 1 row based on the shuffled index** of dataset?

Thanks.

Nic-Ma commented:
`pandas.read_csv()` provides 2 methods to read a CSV file: 1. default to read the whole content. 2. set chunksize and return an iterator.
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html
@ericspod @rijobro @wyli What's the typical use case for a very big CSV file during training?

Thanks in advance.

Nic-Ma commented:
Hi @wyli ,

I am changing this PR to be a MONAI Dataset, could you please help share some typical use cases with very big CSV file?

Thanks in advance.

wyli commented:
> Hi @wyli ,
> 
> I am changing this PR to be a MONAI Dataset, could you please help share some typical use cases with very big CSV file?
> 
> Thanks in advance.

some examples:
for the basic dataset API:
https://github.com/NYUMedML/GNN_for_EHR/blob/master/dataloader.py  and it could be iterabledatasets, something like this text classification example from pytorch: https://github.com/pytorch/text/blob/master/examples/text_classification/iterable_train.py
https://github.com/CogStack/MedCAT/blob/master/medcat/preprocessing/iterators.py

I feel the initial CSV support could be similar to the existing image dataset https://github.com/Project-MONAI/MONAI/blob/82fc4c1b543c2469ef4464a9a60b9a5cd89c67fc/monai/data/image_dataset.py#L23 but with CSV file path as input, with two versions for Dataset and IterableDataset respectively,  support of csv column name/datatype selections

wyli reviewed (changes requested):
had a meeting with Brads and Carole today, this needs some more work.

wyli commented:
> > Hi @wyli ,
> > I am changing this PR to be a MONAI Dataset, could you please help share some typical use cases with very big CSV file?
> > Thanks in advance.
> 
> some examples:
> for the basic dataset API:
> https://github.com/NYUMedML/GNN_for_EHR/blob/master/dataloader.py and it could be iterabledatasets, something like this text classification example from pytorch: https://github.com/pytorch/text/blob/master/examples/text_classification/iterable_train.py
> https://github.com/CogStack/MedCAT/blob/master/medcat/preprocessing/iterators.py
> 
> I feel the initial CSV support could be similar to the existing image dataset
> 
> https://github.com/Project-MONAI/MONAI/blob/82fc4c1b543c2469ef4464a9a60b9a5cd89c67fc/monai/data/image_dataset.py#L23
> 
> but with CSV file path as input, with two versions for Dataset and IterableDataset respectively, support of csv column name/datatype selections

Have you looked at these examples? @Nic-Ma is it possible to follow a similar approach?

Nic-Ma commented:
/black

Nic-Ma commented:
/black

Nic-Ma commented:
Hi @wyli ,

Thanks for providing the examples as use cases, yes, I checked them and updated the PR to have a regular `CSVDataset` and a `CSVIterableDataset`. It can support multiple CSV files, multi-processing and groups.
Could you please help review it again?
Note that the multi-processing implementation of this PR is slightly different from your example, because the example requires the total length of data and then split all data for workers to handle, but I think we should suppose the IterableDataset to be **endless** streaming, so I changed to split every chunk for workers to execute transforms in parallel.

Thanks in advance.

Nic-Ma commented:
/black

Nic-Ma commented:
I didn't figure out why the min_tests of python 3.6 failed, @ericspod @wyli @rijobro do you python experts know something?

Thanks in advance.

ericspod commented:
I'd suggest the assert should be changed to have an error message that states the contents of `err_mod` so we can see what's going on, rather than printing the contents of `err_mod` every time. It looks like MONAI isn't being loaded at all, it would help to detect if this fails and try to import with `import monai` again and use the exception from that to see what's going on.

Nic-Ma commented:
/black

Nic-Ma commented:
/black

Nic-Ma commented:
Hi @ericspod ,

Thanks for your suggestions, I tried to print out error message locally and solved the issue.

@wyli The GPU tests failed due to below error:
```During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/parameterized/parameterized.py", line 533, in standalone_func
    return func(*(a + p.args), **p.kwargs)
  File "/__w/MONAI/MONAI/tests/test_gmm.py", line 323, in test_cuda
    gmm = GaussianMixtureModel(features_tensor.size(1), mixture_count, class_count)
  File "/__w/MONAI/MONAI/monai/networks/layers/gmm.py", line 42, in __init__
    "gmm", {"CHANNEL_COUNT": channel_count, "MIXTURE_COUNT": mixture_count, "MIXTURE_SIZE": mixture_size}
  File "/__w/MONAI/MONAI/monai/_extensions/loader.py", line 91, in load_module
    verbose=verbose_build,
  File "/usr/lib/python3.6/contextlib.py", line 99, in __exit__
    self.gen.throw(type, value, traceback)
  File "/__w/MONAI/MONAI/monai/_extensions/loader.py", line 37, in timeout
    raise TimeoutError(message)
TimeoutError: Build appears to be blocked. Is there a stopped process building the same extension?
```
Should I wait a while and try again?

Thanks.

wyli commented on monai/data/utils.py:
```diff
@@ -983,3 +988,56 @@ def sorted_dict(item, key=None, reverse=False):
+    dfs,
+    row_indices: Optional[Sequence[Union[int, str]]] = None,
+    col_names: Optional[Sequence[str]] = None,
+    col_groups: Optional[Dict[str, Sequence[str]]] = None,
```
we need some basic support for setting defaults for the missing values and also datatypes (as I mentioned in this example https://github.com/Project-MONAI/MONAI/issues/2310#issuecomment-864000564).

Nic-Ma replied:
Hi @wyli ,

Sorry I misunderstood your point yesterday, do you mean to add another arg `column_types` to translate loaded values?
```
column_types = {  # this could come from some json-based mapping file https://www.w3.org/2013/csvw/wiki/CSV-LD)
    "subject_id": {"type": str}, 
    "label": {"type": int, "default": 0},
    "ehr_0": {"type": float, "default": 0.0, "unit": "year"},
    "ehr_1": {"type": float, "default": 0.0, "unit": "month"},
    "image": {"type": nifti, "default": None},
}
```

Nic-Ma replied:
Hi @wyli ,

I added support for `type` and `default`, not sure how to use `unit` to handle data so I didn't add it.
Maybe we can add it later if user requests it.

Thanks.

Nic-Ma commented:
/black

wyli commented:
> Hi @ericspod ,
> 
> Thanks for your suggestions, I tried to print out error message locally and solved the issue.
> 
> @wyli The GPU tests failed due to below error:
> 
> ```
> 
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.6/dist-packages/parameterized/parameterized.py", line 533, in standalone_func
>     return func(*(a + p.args), **p.kwargs)
>   File "/__w/MONAI/MONAI/tests/test_gmm.py", line 323, in test_cuda
>     gmm = GaussianMixtureModel(features_tensor.size(1), mixture_count, class_count)
>   File "/__w/MONAI/MONAI/monai/networks/layers/gmm.py", line 42, in __init__
>     "gmm", {"CHANNEL_COUNT": channel_count, "MIXTURE_COUNT": mixture_count, "MIXTURE_SIZE": mixture_size}
>   File "/__w/MONAI/MONAI/monai/_extensions/loader.py", line 91, in load_module
>     verbose=verbose_build,
>   File "/usr/lib/python3.6/contextlib.py", line 99, in __exit__
>     self.gen.throw(type, value, traceback)
>   File "/__w/MONAI/MONAI/monai/_extensions/loader.py", line 37, in timeout
>     raise TimeoutError(message)
> TimeoutError: Build appears to be blocked. Is there a stopped process building the same extension?
> ```
> 
> Should I wait a while and try again?
> 
> Thanks.

looks like multiple ninja builds sharing the same build cache is still creating some issues. any idea @charliebudd ?

wyli reviewed (approved):
thanks, please add some basic support for missing values, we can have another iteration to update the modules

Nic-Ma commented:
/black
